### ISSUE TITLE
[Investigation] Loading eglGetCurrentDisplay function pointer causes segfault

### ISSUE THREAD

**Issue by j4qfrost** [Kethku/neovide#661]
While debugging the opengl branch, I discovered that it was possible to simply bypass the eglGetCurrentDisplay function to get the branch to run. While this is currently a workable solution, we may run into issues down the road.

From further investigation, I have discovered that glutin loads a glx context by default as opposed to an egl context when using X11. https://github.com/rust-windowing/glutin/blob/master/glutin/src/platform_impl/unix/x11.rs

In order to validate that the context loaded is the main cause we would need to test the function loading on a machine that uses Wayland, since glutin's implementation for Wayland defaults to egl. https://github.com/rust-windowing/glutin/blob/master/glutin/src/platform_impl/unix/wayland.rs

After the issue is confirmed, another issue should be raised with glutin to discuss possible implementation changes, or whether was should simply continue on ignoring eglGetCurrentDisplay for x11 systems. An issue should also be raised with https://github.com/rust-skia/rust-skia, since there is no test or example that catches this issue.

One curious problem that may point away from my theory that this is caused by the loading of the wrong context is that 
eglQueryString still loads fine, so the correct context may be loaded, but the issue just lies somewhere else.

**Comment by j4qfrost** [comment]
TL;DR help wanted from Wayland users. Please test.

**Comment by p00f** [comment]
What exactly should I do to test?

**Comment by j4qfrost** [comment]
@p00f you can either use an older version of the opengl branch from before the egl patch was added or comment out the if statement for eglGetCurrentDisplay

**Comment by p00f** [comment]
I set a breakpoint here:

https://github.com/Kethku/neovide/blob/3ba98d05d113c81412be7a6f52bfed5616c21075/src/window/window_wrapper/renderer.rs#L54

On stepping over, it jumped directly to:

https://github.com/Kethku/neovide/blob/3ba98d05d113c81412be7a6f52bfed5616c21075/src/window/window_wrapper/renderer.rs#L57

So I think it isn't loaded

**Comment by p00f** [comment]
uh huh this happens even when i run it under xwayland

**Comment by p00f** [comment]
I added debug statements
diff --git a/src/window/window_wrapper/renderer.rs b/src/window/window_wrapper/renderer.rs
index 9d2a8aa..78f4c84 100644
--- a/src/window/window_wrapper/renderer.rs
+++ b/src/window/window_wrapper/renderer.rs
@@ -1,3 +1,4 @@
+use log::debug;
 use skia_safe::gpu::gl::FramebufferInfo;
 use skia_safe::gpu::{BackendRenderTarget, DirectContext, SurfaceOrigin};
 use skia_safe::{Canvas, ColorType, Surface};
@@ -51,6 +52,7 @@ impl SkiaRenderer {
         gl::load_with(|s| windowed_context.get_proc_address(&s));
 
         let interface = skia_safe::gpu::gl::Interface::new_load_with(|name| {
+            eprintln!("{}", name);
             if name == "eglGetCurrentDisplay" {
                 return std::ptr::null();
             }

**Comment by p00f** [comment]
the logs are exactly the same in both wayland and xwayland: https://p.ip.fi/33Zu

**Comment by p00f** [comment]
when i set a breakpoint at

https://github.com/Kethku/neovide/blob/3ba98d05d113c81412be7a6f52bfed5616c21075/src/window/window_wrapper/renderer.rs#L54

and step over, it skips the `if name == "eglGetCurrentDisplay"` block in both wayland and xwayland and jumps to

https://github.com/Kethku/neovide/blob/3ba98d05d113c81412be7a6f52bfed5616c21075/src/window/window_wrapper/renderer.rs#L57

**Comment by p00f** [comment]
and me running under wayland vs xwayland is indeed different, because when i run under pure wayland i get the known bug - swapping keys in WM settings doesn't swap them in neovide

**Comment by p00f** [comment]
Both wayland and xwayland work correctly for me, both with and without commenting the block out

**Comment by p00f** [comment]
https://github.com/Kethku/neovide/blob/3ba98d05d113c81412be7a6f52bfed5616c21075/src/window/window_wrapper/renderer.rs#L54 you're talking about this right?

**Comment by Kethku** [comment]
Thats the one. I'm sorta at a loss for why it works on some setups but not others. This was my experience as well...

**Comment by Kethku** [comment]
The hack seems to work pretty consistently. Closing as fixed.

**Comment by Kethku** [comment]
@j4qfrost feel free to reopen if you think we should investigate further.

**Comment by j4qfrost** [comment]
I think it's more appropriate to open an issue with glutin. I dove into the source and it behaves in the way that I suspected. I'll mention this issue after I make a proper issue in glutin.
